### ISSUE TITLE
Downloading pages in parallel does not work?

### ISSUE THREAD

**Issue by TheHardew** [tachiyomiorg/tachiyomi#7258]
### Steps to reproduce

Using the v0.13.4 tag (I tried building 4560033e6 but couldn't):
- Use this patch:
```
diff --git a/app/src/main/java/eu/kanade/tachiyomi/data/download/Downloader.kt b/app/src/main/java/eu/kanade/tachiyomi/data/download/Downloader.kt
index e60046be5..d953ba3b8 100644
--- a/app/src/main/java/eu/kanade/tachiyomi/data/download/Downloader.kt
+++ b/app/src/main/java/eu/kanade/tachiyomi/data/download/Downloader.kt
@@ -39,6 +39,8 @@ import java.io.File
 import java.util.zip.CRC32
 import java.util.zip.ZipEntry
 import java.util.zip.ZipOutputStream
+import java.lang.Thread
+import android.util.Log
 
 /**
  * This class is the one in charge of downloading chapters.
@@ -62,6 +64,7 @@ class Downloader(
 ) {
 
     private val chapterCache: ChapterCache by injectLazy()
+	private var conI: Int = 0;
 
     private val preferences: PreferencesHelper by injectLazy()
 
@@ -365,6 +368,8 @@ class Downloader(
      */
     private fun getOrDownloadImage(page: Page, download: Download, tmpDir: UniFile): Observable<Page> {
         // If the image URL is empty, do nothing
+		conI++
+		Log.i("DownloadTachiyomiImage", "Entering: " + conI)
         if (page.imageUrl == null) {
             return Observable.just(page)
         }
@@ -392,12 +397,16 @@ class Downloader(
                 page.progress = 100
                 download.downloadedImages++
                 page.status = Page.READY
+				conI--
+				Log.i("DownloadTachiyomiImage", "Exiting: " + conI)
             }
             .map { page }
             // Mark this page as error and allow to download the remaining
             .onErrorReturn {
                 page.progress = 0
                 page.status = Page.ERROR
+				conI--
+				Log.i("DownloadTachiyomiImage", "Exiting: " + conI)
                 page
             }
     }
@@ -413,6 +422,7 @@ class Downloader(
     private fun downloadImage(page: Page, source: HttpSource, tmpDir: UniFile, filename: String): Observable<UniFile> {
         page.status = Page.DOWNLOAD_IMAGE
         page.progress = 0
+		Thread.sleep(2_000)
         return source.fetchImage(page)
             .map { response ->
                 val file = tmpDir.createFile("$filename.tmp")

```
- install the app
- Open tachiyomi, download some manga chapter (the application might crash, open it again, if a wheel is spinning next to the chapter, press on it and start the download now, if not, try again from the beginning).
- `adb logcat | fgrep DownloadTachiyomiImage`

### Expected behavior

The pages should be downloaded 5 at a time. From the "Actulal behavior" section you can see they are not.
If you use this patch instead (which also subscribes the function that downloads the image on an io thread):
```
diff --git a/app/src/main/java/eu/kanade/tachiyomi/data/download/Downloader.kt b/app/src/main/java/eu/kanade/tachiyomi/data/download/Downloader.kt
index e60046be5..f71129bee 100644
--- a/app/src/main/java/eu/kanade/tachiyomi/data/download/Downloader.kt
+++ b/app/src/main/java/eu/kanade/tachiyomi/data/download/Downloader.kt
@@ -39,6 +39,8 @@ import java.io.File
 import java.util.zip.CRC32
 import java.util.zip.ZipEntry
 import java.util.zip.ZipOutputStream
+import java.lang.Thread
+import android.util.Log
 
 /**
  * This class is the one in charge of downloading chapters.
@@ -62,6 +64,7 @@ class Downloader(
 ) {
 
     private val chapterCache: ChapterCache by injectLazy()
+	private var conI: Int = 0;
 
     private val preferences: PreferencesHelper by injectLazy()
 
@@ -339,7 +342,7 @@ class Downloader(
             .flatMap { download.source.fetchAllImageUrlsFromPageList(it) }
             // Start downloading images, consider we can have downloaded images already
             // Concurrently do 5 pages at a time
-            .flatMap({ page -> getOrDownloadImage(page, download, tmpDir) }, 5)
+            .flatMap({ page -> getOrDownloadImage(page, download, tmpDir).subscribeOn(Schedulers.io()) }, 5)
             .onBackpressureLatest()
             // Do when page is downloaded.
             .doOnNext { notifier.onProgressChange(download) }
@@ -365,6 +368,8 @@ class Downloader(
      */
     private fun getOrDownloadImage(page: Page, download: Download, tmpDir: UniFile): Observable<Page> {
         // If the image URL is empty, do nothing
+		conI++
+		Log.i("DownloadTachiyomiImage", "Entering: " + conI)
         if (page.imageUrl == null) {
             return Observable.just(page)
         }
@@ -392,12 +397,16 @@ class Downloader(
                 page.progress = 100
                 download.downloadedImages++
                 page.status = Page.READY
+				conI--
+				Log.i("DownloadTachiyomiImage", "Exiting: " + conI)
             }
             .map { page }
             // Mark this page as error and allow to download the remaining
             .onErrorReturn {
                 page.progress = 0
                 page.status = Page.ERROR
+				conI--
+				Log.i("DownloadTachiyomiImage", "Exiting: " + conI)
                 page
             }
     }
@@ -413,6 +422,7 @@ class Downloader(
     private fun downloadImage(page: Page, source: HttpSource, tmpDir: UniFile, filename: String): Observable<UniFile> {
         page.status = Page.DOWNLOAD_IMAGE
         page.progress = 0
+		Thread.sleep(2_000)
         return source.fetchImage(page)
             .map { response ->
                 val file = tmpDir.createFile("$filename.tmp")

```

the output will be:
```
06-05 23:12:53.658 13436 13700 I DownloadTachiyomiImage: Entering: 1
06-05 23:12:55.672 13436 13700 I DownloadTachiyomiImage: Entering: 2
06-05 23:12:57.682 13436 13700 I DownloadTachiyomiImage: Entering: 3
06-05 23:12:59.034 13436 13701 I DownloadTachiyomiImage: Exiting: 2
06-05 23:12:59.057 13436 13696 I DownloadTachiyomiImage: Exiting: 1
06-05 23:12:59.696 13436 13700 I DownloadTachiyomiImage: Entering: 2
06-05 23:12:59.940 13436 13699 I DownloadTachiyomiImage: Exiting: 1
06-05 23:13:01.706 13436 13700 I DownloadTachiyomiImage: Entering: 2
06-05 23:13:01.920 13436 13696 I DownloadTachiyomiImage: Exiting: 1
06-05 23:13:03.719 13436 13700 I DownloadTachiyomiImage: Entering: 2
06-05 23:13:04.116 13436 13701 I DownloadTachiyomiImage: Exiting: 1
06-05 23:13:05.730 13436 13700 I DownloadTachiyomiImage: Entering: 2
06-05 23:13:05.935 13436 13699 I DownloadTachiyomiImage: Exiting: 1
06-05 23:13:07.743 13436 13700 I DownloadTachiyomiImage: Entering: 2
06-05 23:13:08.105 13436 13696 I DownloadTachiyomiImage: Exiting: 1
06-05 23:13:09.756 13436 13700 I DownloadTachiyomiImage: Entering: 2
06-05 23:13:10.018 13436 13701 I DownloadTachiyomiImage: Exiting: 1
```

### Actual behavior

Only one image is downloaded at a time:
```
06-05 23:06:18.366 10525 11124 I DownloadTachiyomiImage: Entering: 1
06-05 23:06:24.029 10525 11124 I DownloadTachiyomiImage: Exiting: 0
06-05 23:06:24.057 10525 11124 I DownloadTachiyomiImage: Entering: 1
06-05 23:06:26.456 10525 11124 I DownloadTachiyomiImage: Exiting: 0
06-05 23:06:26.465 10525 11124 I DownloadTachiyomiImage: Entering: 1
06-05 23:06:28.836 10525 11124 I DownloadTachiyomiImage: Exiting: 0
06-05 23:06:28.845 10525 11124 I DownloadTachiyomiImage: Entering: 1
06-05 23:06:31.049 10525 11124 I DownloadTachiyomiImage: Exiting: 0
06-05 23:06:31.065 10525 11124 I DownloadTachiyomiImage: Entering: 1
06-05 23:06:33.219 10525 11124 I DownloadTachiyomiImage: Exiting: 0
06-05 23:06:33.237 10525 11124 I DownloadTachiyomiImage: Entering: 1
06-05 23:06:35.512 10525 11124 I DownloadTachiyomiImage: Exiting: 0
06-05 23:06:35.523 10525 11124 I DownloadTachiyomiImage: Entering: 1
06-05 23:06:38.052 10525 11124 I DownloadTachiyomiImage: Exiting: 0
06-05 23:06:38.066 10525 11124 I DownloadTachiyomiImage: Entering: 1
06-05 23:06:40.323 10525 11124 I DownloadTachiyomiImage: Exiting: 0
06-05 23:06:40.333 10525 11124 I DownloadTachiyomiImage: Entering: 1
06-05 23:06:42.636 10525 11124 I DownloadTachiyomiImage: Exiting: 0
```

### Tachiyomi version

0.13.4

### Android version

Android 11

### Device

Samsung Galaxy S21+

### Other details

I'm opening this issue as suggested in the previous one: https://github.com/tachiyomiorg/tachiyomi-extensions/issues/12025

Even with the `.subscribeOn(...)` you can see, that it reaches 3 images being downloaded at once at the beginning and then it oscillates between 1 and 2, it does not reach 5, no idea why that is.

Even without any code modifications, when monitoring tcp connections on my Komga server using `ss -s` from `iproute2`, the app only ever tried to download 1 page at a time, so I don't think the added sleep or logging is the cause of the problem.

I don't know kotlin/ReactiveX/java

### Acknowledgements

- [X] I have searched the existing issues and this is a new ticket, **NOT** a duplicate or related to another open issue.
- [X] I have written a short but informative title.
- [X] If this is an issue with an extension, I should be opening an issue in the [extensions repository](https://github.com/tachiyomiorg/tachiyomi-extensions/issues/new/choose).
- [X] I have tried the [troubleshooting guide](https://tachiyomi.org/help/guides/troubleshooting/).
- [X] I have updated the app to version **[0.13.4](https://github.com/tachiyomiorg/tachiyomi/releases/latest)**.
- [X] I have updated all installed extensions.
- [X] I will fill out all of the requested information in this form.

**Comment by arkon** [comment]
Closing in favour of #7492
